shell/platform/glfw/key_event_handler.cc

```diff
@@ -12,14 +12,96 @@ static constexpr char kChannelName[] = "flutter/keyevent";
 
 static constexpr char kKeyCodeKey[] = "keyCode";
 static constexpr char kKeyMapKey[] = "keymap";
+static constexpr char kScanCodeKey[] = "scanCode";
+static constexpr char kModifiersKey[] = "modifiers";
 static constexpr char kTypeKey[] = "type";
+static constexpr char kToolkitKey[] = "toolkit";
+static constexpr char kUnicodeScalarValuesProduced[] =
+    "unicodeScalarValuesProduced";
+
+static constexpr char kLinuxKeyMap[] = "linux";
+static constexpr char kGLFWKey[] = "glfw";
 
-static constexpr char kAndroidKeyMap[] = "android";
 static constexpr char kKeyUp[] = "keyup";
 static constexpr char kKeyDown[] = "keydown";
 
+// Masks used for UTF-8 to UTF-32 conversion.
+static constexpr int kTwoByteMask = 0xC0;
+static constexpr int kThreeByteMask = 0xE0;
+static constexpr int kFourByteMask = 0xF0;
+
 namespace flutter {
 
+namespace {
+
+// Information about the UTF-8 encoded code point.
+struct UTF8CodePointInfo {
+  // The bit-mask that determines the length of the code point.
+  int first_byte_mask;
+  // The number of bytes of the code point.
+  size_t length;
+};
+
+// Creates a [UTF8CodePointInfo] from a given byte. [first_byte] must be the
+// first byte in the code point.
+UTF8CodePointInfo GetUTF8CodePointInfo(int first_byte) {
+  UTF8CodePointInfo byte_info;
+
+  // The order matters. Otherwise, it is possible that comparing against i.e.
+  // kThreeByteMask and kFourByteMask could be both true.
+  if ((first_byte & kFourByteMask) == kFourByteMask) {
+    byte_info.first_byte_mask = 0x07;
+    byte_info.length = 4;
+  } else if ((first_byte & kThreeByteMask) == kThreeByteMask) {
+    byte_info.first_byte_mask = 0x0F;
+    byte_info.length = 3;
+  } else if ((first_byte & kTwoByteMask) == kTwoByteMask) {
+    byte_info.first_byte_mask = 0x1F;
+    byte_info.length = 2;
+  } else {
+    byte_info.first_byte_mask = 0xFF;
+    byte_info.length = 1;
+  }
+  return byte_info;
+}
+
+// Queries GLFW for the printable key name given a [key] and [scan_code] and
+// converts it to UTF-32. The Flutter framework accepts only one code point,
+// therefore, only the first code point will be used. This unlikely but there is
```

Comment:
s/This is unlikely/There is unlikely to be more than one/ (the "This" currently doesn't refer to anything that was actually said before.)

Reply:
done